tutorials/image/cifar10_estimator/cifar10_main.py

```diff
@@ -220,147 +145,172 @@ def __call__(self, op):
     return device_name
 
 
-def _create_device_setter(is_cpu_ps, worker, num_gpus):
+def _create_device_setter(avg_on_gpu, worker, num_gpus):
```

Comment:
RE: `--avg-on-gpu` am I correct in noting that there are two separate things being controlled by this flag?

(1) whether variables are placed on GPU or not.
(2) Whether the averaging ops are placed on GPU or not.

Note for (2) I'm still a little confused, because there appear to be two different device setters, one at 236, and one that uses `var.device` at 243. If I understand device contexts properly using explicit devices rather than device functions makes the implicit device function `merge_devices`, which is conservative. So the averaging will always happen on the device the variables are located on (which makes sense, so that the final gradients are colocated with the device on which the update op takes place). So I'm not quite sure what the outer device context is doing? Maybe a few concats and an average of loss? But I think it would be clearer if they weren't nested. Given that the variable devices are going to be fully specified, the outer context will always be overridden by the inner context.